### DISCUSSION TITLE
PutBucketLifecycle(deprecated) vs PutBucketLifecycleConfiguration

### DISCUSSION THREAD

**Discussion by vlvicente** [minio/minio#20580] [Q&A] [answered]
Can anyone confirm that the new version of LifecycleConfiguration is not yet supported by MinIO?
Context: I've been trying to programmatically add a lifecycle configuration to a bucket, which is rejected with a bad request saying the XML is not wellformed. I'm using [rust-s3](https://docs.rs/rust-s3/latest/s3/index.html) which only implements the new version and the xml does seem well formed and within the specs defined by AWS.
Taking a look at the MinIO implementation I noticed that one of the validations made is this one:
```go
func (r Rule) validatePrefixAndFilter() error {
 if !r.Prefix.set && r.Filter.IsEmpty() || r.Prefix.set && !r.Filter.IsEmpty() {
   return errXMLNotWellFormed
 }
 if !r.Prefix.set {
   return r.Filter.Validate()
 }
 return nil
}
```
which seems to consider the [deprecated version](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycle.html) of the request (in which the Prefix field is in the same level as Filter and marked as required) instead of [new version](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycleConfiguration.html) (Prefix is moved inside the Filter field)

**Comment by dhananjaykrutika** [comment]
@vlvicente do you mind sharing the rule you're trying to add?

**Reply by vlvicente** [reply]
The rule I'm trying to add is a simple Expiration. Using the PutBucketLifecycleConfiguration spec instead of the deprecated PutBucketLifecycle. 
The request body looks like this:
```xml
<LifecycleConfiguration>
    <Rule>
        <Expiration>
            <Days>6</Days>
        </Expiration>
        <ID>ExpireRule</ID>
    </Rule>
</LifecycleConfiguration>
```
which does not work on the minio deployment.
If we add the `Prefix` field, it works (Deprecated version):
```xml
<LifecycleConfiguration>
    <Rule>
        <Expiration>
            <Days>6</Days>
        </Expiration>
        <ID>ExpireRule</ID>
        <Filter>
            <Prefix></Prefix>
        </Filter>
        <Status>Enabled</Status>
    </Rule>
</LifecycleConfiguration>
```

The deprecated [PutBucketLifecycle](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycle.html) uses a Rule object which requires the fields `Prefix` or `Filter`. the (new version)[https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycleConfiguration.html] do not require these fields.

**Comment by dhananjaykrutika** [comment, marked as answer]
Here is a potential fix to this problem - https://github.com/minio/minio/pull/20684
